### PR TITLE
PR: Improve performance when processing linting results and scrolling (Editor)

### LINKED ISSUE #13020: Editor sluggish on document diagnostics update
### Issue Report Checklist

* [x ] Searched the [issues page](https://github.com/spyder-ide/spyder/issues?q=is%3Aissue) for similar reports
* [ x] Read the relevant sections of the [Spyder Troubleshooting Guide](https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ) and followed its advice
* [ x] Reproduced the issue after updating with ``conda update spyder`` (or ``pip``, if not using Anaconda)

## Problem Description
On mac (macOS 10.15.5, in this case), the editor in spyder seems to be laggy and slow. Simple tests are fast typing - sometimes it takes seconds to fill all characters. Scrolling is also slow - the editor keeps on rolling for seconds after the wheel on the mouse stops until the "buffer" is worked on...
Autocomplete needs up to seconds to react. However, it does not seem to be autocorrect related since disable alls features in this class do not change anything - neither does styles like vertical lines...
All tests are on a 4k screen, different scalings tested. The UI seems to be responsive anyway.

### What steps reproduce the problem?

1. Start spyder on macOS and start typing in the editor (**!not the console - that works fast!**)

### Paste Traceback/Error Below (if applicable)

```python-traceback

PASTE TRACEBACK HERE

```

## Versions
* Spyder version: 4.1.3
* Python version: 3.7.7
* Qt version: 5.12.8
* PyQt version: 5.12.3
* Operating System name/version: macOS10.15.5 (Darwin 19.4.0) 

### Dependencies
```

# Mandatory:
applaunchservices >=0.1.7      :  0.2.1 (OK)
atomicwrites >=1.2.0           :  1.4.0 (OK)
chardet >=2.0.0                :  3.0.4 (OK)
cloudpickle >=0.5.0            :  1.4.1 (OK)
diff_match_patch >=20181111    :  20181111 (OK)
intervaltree                   :  None (OK)
IPython >=4.0                  :  7.15.0 (OK)
jedi =0.15.2                   :  0.15.2 (OK)
nbconvert >=4.0                :  5.6.1 (OK)
numpydoc >=0.6.0               :  1.0.0 (OK)
parso =0.5.2                   :  0.5.2 (OK)
pexpect >=4.4.0                :  4.8.0 (OK)
pickleshare >=0.4              :  0.7.5 (OK)
psutil >=5.3                   :  5.7.0 (OK)
pygments >=2.0                 :  2.6.1 (OK)
pylint >=0.25                  :  2.5.3 (OK)
pyls >=0.31.9;<0.32.0          :  0.31.10 (OK)
qdarkstyle >=2.8               :  2.8.1 (OK)
qtawesome >=0.5.7              :  0.7.2 (OK)
qtconsole >=4.6.0              :  4.7.4 (OK)
qtpy >=1.5.0                   :  1.9.0 (OK)
sphinx >=0.6.6                 :  3.1.0 (OK)
spyder_kernels >=1.9.1;<1.10.0 :  1.9.1 (OK)
watchdog                       :  None (OK)
zmq >=17                       :  19.0.1 (OK)

# Optional:
cython >=0.21                  :  None (OK)
matplotlib >=2.0.0             :  3.2.1 (OK)
numpy >=1.7                    :  1.18.5 (OK)
pandas >=0.13.1                :  1.0.4 (OK)
scipy >=0.17.0                 :  1.4.1 (OK)
sympy >=0.7.3                  :  None (OK)
```

### PR DESCRIPTION
**Pull request by ccordoba12**
This PR does the following:

* Break how we are processing linting results into two subprocesses:
    * Set block user data in a thread. That helps a lot when processing too many linting results.
    * Underline linting results only in the visible portion of the editor. This also helps performance.
* Update decorations after a small timeout. This prevents multiple updates being called almost at the same time.
* Paint all decorations only in the visible portion of the editor.
* Remove repeated calls to `highlight_current_cell`.
* Add buffer around the visible portion of the screen to give a more fluid experience when updating decorations.
* Fix scrolling sluggishness in split editors without focus.

Fixes #13020.

### CODE REVIEW

**Comment by ccordoba12** [comment]
@dalthviz, please help me to test if this solves the sluggishness reported on issue #13020.

@CAM-Gerlach, please help me to check if this solves #8864.

**Comment by CAM-Gerlach** [comment]
@ccordoba12 Sorry for the delay, testing in work now.

**Comment by dalthviz** [comment]
@ccordoba12 it seems like the underline error performance improved (although as @CAM-Gerlach says there is now a little delay when scrolling to see again the underlines)

**Comment by ccordoba12** [comment]
>  it seems like the underline error performance improved

How much has it improved? I mean, would you way issue #13020 is fixed by this PR?

**Comment by CAM-Gerlach** [comment]
Hey @dalthviz can you suggest an objective test I can perform to quantify this, besides the above? The results of my benchmarks indicated around -10 to +20% change, or around ~5% average (since most tests showed no change), in basic text editing responsiveness (the problem cited in #13020 ), on a 1200p monitor with a margin of error of ~10-20%. for individual tests and ~5% average. In particular, typing speed (the main objective metric cited in #13020 ) regressed by 10% under default settings, improved by 7% with defaults + indent guides, and showed no change with all decorations enabled minus indent guides, which is all within the margin of error for the tests and likely not significant.

**Comment by CAM-Gerlach** [comment]
@max3-2 and @mrclary would you like to test this, either with similar benchmarks as I execute above or with those that better reflect the specific issues you're seeing? Thanks!

**Comment by dalthviz** [comment]
@CAM-Gerlach probably doing a profiling could be helpful using the same context as explained here: https://github.com/spyder-ide/spyder/issues/13020#issuecomment-654848318 by @colinww

**Comment by mrclary** [comment]
> @max3-2 and @mrclary would you like to test this, either with similar benchmarks as I execute above or with those that better reflect the specific issues you're seeing? Thanks!

I'll take a look...

**Comment by colinww** [comment]
I just spent a few minutes browsing the same file as before `lib/python3.7/site-packages/CppHeaderParser/CppHeaderParser.py` and I can confirm the UI is night and day better. Scrolling is faster, cursor lag is basically gone, and the file has a ton of linting errors that don't seem to slow the editor down at all. On the release version, drawing the under squigglies seemed to be the main culprit for slowdown.

When I cloned the repo I did notice it updated my python to 3.8.3, and ipython as well as a dozen other packages. So I can't confirm exactly this PR is responsible for the whole delta but it is so much better I am just going to start doing my normal work in this dev environment.

**Comment by CAM-Gerlach** [comment]
@colinww Thanks for the feedback! Do you think you could to the `4.x` branch and this PR to confirm the difference? Launching Spyder from a development branch uses a different set of settings, and there have been other recent changes targeted at improving performance, so that would help isolate the impact of this one.

I'm not sure why I'm not seeing it in any of my benchmarks—what monitor size, OS version and CPU/machine are you using? Any ideas for something else I can test that will expose the difference?

@ccordoba12 Maybe I'd see more of a difference if I tested with just the decorations this PR affects, e.g. if I enable the code style linting and underlines but not showing blank spaces (which introduce considerable lag of their own)? Any tips here?

**Comment by colinww** [comment]
I have a work deadline so I will need a day or two to try this out.

The machine I am on now is a 16" MBP, it has a 8 core i9, 64GB RAM, 3840x1600 screen. In our office, we run spyder on servers running Centos 6, Thinkpad laptops running windows 10, and the same laptops running the linux version of spyder through WSL/Ubuntu. In terms of this editor/GUI lag, I think the worst is on native Centos 6, then probably OSX, then WSL/Ubuntu, then Windows. In Centos 6, even though our servers are 64 core with 2TB RAM, it was so unusable we actually setup samba mounts so people could edit their files through windows version of spyder. For some reason, it just doesn't seem to be as bad on Windows.

**Comment by max3-2** [comment]
Thanks for the improvement! I'll give it a test until the weekend and report back, won't make it any earlier sorry.

**Comment by max3-2** [comment]
I did some "not-benchmarking" tests with some larger files and while it is better, it still seems to be there, again especially on larger files >1500 lines. I have the snakeviz

 profiling attached. The remove seems to be gone though. I tested with decorations off and on...
[spyder.profile.zip](https://github.com/spyder-ide/spyder/files/4946205/spyder.profile.zip)

**Comment by mrclary** [comment]
* I did not observe an improvement for typing in the editor.
* I did not observe an improvement for scrolling in the editor.

Attached are my profiles with and without indent guides enabled. 
* with IG: indentationguides.py paintEvent is prominent
* without IG: editor.py focus_changed and refresh are prominent

I have discovered an interesting phenomenon.
With IG on and window split vertically, both top and bottom splits' scrolling are sluggish. However, with IG off and the editor window split vertically, the top split scrolling is still sluggish while the bottom split scrolling is snappy (?!). Attached gif illustrates. 

Screen Shot 2020-07-20 at 7 38 56 AM
Screen Shot 2020-07-20 at 7 49 20 AM
Screen Recording 2020-07-20 at 7 44 41 AM mov

[spyder-profiles.zip](https://github.com/spyder-ide/spyder/files/4948436/spyder-profiles.zip)

**Comment by mrclary** [comment]
Without IG

**Comment by max3-2** [comment]
> I have discovered an interesting phenomenon.
> With IG on and window split vertically, both top and bottom splits' scrolling are sluggish. However, with IG off and the editor window split vertically, the top split scrolling is still sluggish while the bottom split scrolling is snappy (?!). Attached gif illustrates.
> 

I can reproduce. However its the smaller window which is snappier with scrolling. I can have 3 or more and its still improving. However when using a number of windows and resizing one to be almost as large as having one, its back to slow again.
Also, **this only affects scrolling, typing is still slow!**

**Comment by mrclary** [comment]
I've noticed something else that may be informative...
Sluggish scrolling seems to go away as soon as there is any interaction with the scrollbar or scrollflag.
When I move the scrollbar or scrollflag with the mouse pointer, even a little, then try scrolling again with swiping, it is real snappy.
When using a split window, this causes the other split to become sluggish, until I diddle the scrollbar.
This behavior manifests independent of code-folding or indent guides.

**Comment by mrclary** [comment]
> I've noticed something else that may be informative...
> Sluggish scrolling seems to go away as soon as there is any interaction with the scrollbar or scrollflag.
> When I move the scrollbar or scrollflag with the mouse pointer, even a little, then try scrolling again with swiping, it is real snappy.
> When using a split window, this causes the other split to become sluggish, until I diddle the scrollbar.
> This behavior manifests independent of code-folding or indent guides.
> 
> There is no affect on typing speed.

The above statement regarding the scrollbar and scrollflag is not entirely accurate. I've discovered that whichever window has _cursor focus_ scrolls smoothly. This is in contrast to mouse focus. I cannot speak for other platforms, but on MacOS, scrolling is possible when the mouse is hovering over a window, regardless of where the typing cursor has focus. If the typing cursor has focus on the window which is scrolled, it scrolls smooth. If the typing cursor has focus elsewhere (Console or other split Editor) then the scrolling will be sluggish. Interacting with the scrollbar or scrollflag actually brings the typing cursor focus to that window, which is why I thought they were related to the issue.

Perhaps this will help further our search for a solution.

**Comment by ccordoba12** [comment]
> I cannot speak for other platforms, but on MacOS, scrolling is possible when the mouse is hovering over a window, regardless of where the typing cursor has focus.

@mrclary, thanks a lot for your feedback! I think I solved this issue in my last commit. Could you try it and let me know if it works for you too?

**Comment by mrclary** [comment]
I’m on vacation right now, but will take a look as soon as I can. 

Thanks,
Ryan Clary

**Comment by mrclary** [comment]
@ccordoba12, sorry for the late response.
I still see sluggish scrolling with your latest commits. I'm not sure why it works smoothly when the text cursor is in the editor, but sluggish when it isn't.

**Comment by mrclary** [comment]
I also just discovered another idiosyncrasy which may help isolate the issue.
Recall that when the Editor is docked, regardless whether the Editor is split or not, any Editor pane without type cursor focus would scroll sluggishly.
However, if the Editor is _undocked_ then the behavior is changed in the following way.
* If the Editor window does not have focus, any Editor pane will scroll smoothly.
* If the Editor has focus, then only the Editor pane with type cursor focus will scroll smoothly.

This is true for previous Spyder commits as well as the latest commits in this PR.

**Comment by ccordoba12** [comment]
@mrclary, I think I finally fixed the sluggishness when scrolling in split editors. Please try again.

File: spyder/plugins/editor/widgets/codeeditor.py
Comment (dalthviz) [inline spyder/plugins/editor/widgets/codeeditor.py:103]: As with `UPDATE_TIMEOUT` should this value be configurable in some way? (Maybe a different value is better for different machines)?

**Review by dalthviz — COMMENTED** [review]
@ccordoba12 left some comments (regarding the constant for the timers). Besides that LGTM :+1:

**Comment by mrclary** [comment]
> @mrclary, I think I finally fixed the sluggishness when scrolling in split editors. Please try again.

SHAZAAM! Great work @ccordoba12! I can no longer reproduce sluggish scrolling under any circumstance!

File: spyder/plugins/editor/utils/decoration.py
Comment (CAM-Gerlach) [reply spyder/plugins/editor/utils/decoration.py:27]: You mean by the user? Is there a way the average user is likely to know how to tune this parameter?

**Comment by CAM-Gerlach** [comment]
@ccordoba12 I can test as well to see if, and how much, it makes a difference with the other metrics I report above.

File: spyder/plugins/editor/utils/decoration.py
Comment (dalthviz) [reply spyder/plugins/editor/utils/decoration.py:27]: Yep, and probably an advanced user or an average user with some guidance could tune it. The thing is that I'm unsure if that value will be suitable in any kind of setup. Maybe referencing here why this value is good for the majority of users or the average case should be enough.

Comment (CAM-Gerlach) [reply spyder/plugins/editor/utils/decoration.py:27]: I'm definitely the type of user to spend time tuning this, but if there isn't an easily intuitive way to determine what value makes sense besides just lots of careful testing, I'm not sure how many other users would be willing to spend the time benchmarking their system unless there's an automated way to do it. However, given how critical this is for Spyder's overall user experience, maybe we could get at least a few of the devs and users to test a range of values on their systems and report which result in the best responsiveness (preferably quantitative), but if this default already gives good results then I'm not sure there's a reason to expose it to the user.

Comment (ccordoba12) [reply spyder/plugins/editor/utils/decoration.py:29]: @dalthviz, I think allowing users to set these values could produce terrible results. For instance, if they make this constant too small (5 or 1 miliseconds), they'll get basically the same behavior they're observing in the editor right now (i.e. a lot of sluggishness). And if they make it too large (20 miliseconds or more), they wouldn't see decorations to be updated promptly (e.g. the current line wouldn't be highlighted immediately as you move the cursor up and down, but you'll see a delay in doing that).

And similar results would be seen for UPDATE_DECORATIONS_TIMEOUT: small values would produce sluggishness and large ones would produce unpleasant delays (as if Spyder were not working).

However, you and @CAM-Gerlach are welcome to play with those constants to see if you can find values more suited for Windows (I've only tested this PR on macOS and Linux).

Comment (ccordoba12) [reply spyder/plugins/editor/utils/decoration.py:29]: Besides, I decided to add comments above this constant and UPDATE_DECORATIONS_TIMEOUT to explain what exactly they are used for (please see my last commits).

**Comment by ccordoba12** [comment]
> @ccordoba12 I can test as well to see if, and how much, it makes a difference with the other metrics I report above.

Thanks for the offer @CAM-Gerlach, but I think @mrclary and @colinww have confirmed this is working for the issues they've observed.

I took a look at our indent guides and found they are updated using a different mechanism than decorations. So I need to open a new PR for that (to not make this one larger than it already is). Then, I'll give you a ping to help me test that one.

**Comment by CAM-Gerlach** [comment]
Okay, great. In my quick testing, as before I didn't observe a significant difference between this branch and `4.x` in the major responsiveness metrics I tested, but I presume those aren't the target here.

There appeared to be a substantial regression in performance across both branches since my last test (everything is about 30-40% slower, similar to or greater than the difference between having no vs. all annotation features enabled minus indent guides), either due to updated dependency versions, common changes between branches, differences in my system resource utilization/background tasks, etc., but this doesn't appear to be related to this PR.

File: spyder/plugins/editor/utils/decoration.py
Comment (CAM-Gerlach) [reply spyder/plugins/editor/utils/decoration.py:29]: Okay, will do. @mrclary @max3-2 @colinww , we'd love your help testing this as well; the more representative variety of machines, OSes and environments we can test on to find a good balance, the better.

**Comment by CAM-Gerlach** [comment]
I ran some additional tests targeted at the specific indicators @ccordoba12, specifically related to scrolling speed with and without decorations (style linting and underlining) enabled. All tests were done in `spyder/app/mainwindow.py` as before (a ~4000 line file with hundreds of style warnings), around line 1000 for the non-full document tests.

Scrolling 100 lines with arrow keys, with the start line initially positioned at the edge of the window in the scroll direction:
No decorations, 4.x: 14.5-15 lines/s (down), 14.5-15 lines/s (up)
No decorations, this PR: 15-15.5 lines/s (down), 114.5-15 lines/s (up)

Decorations, 4.x: 13.5-14 lines/s (down); 13-13.5 lines/s (up)
Decorations, this PR: 14-14.5 lines/s (down); 13-13.5 lines/s (up)


Scrolling 3675 lines (full file) with pageup/pagedown keys:
No decorations, 4.x: 1050-1100 lines/s (down), 1050-1150 lines/s (up)
No decorations, this PR: 1050-1100 lines/s (down), 1050-1150 lines/s (up)

Decorations, 4.x: 700-900 lines/s (down); 700-1000 lines/s (up)
Decorations, this PR: 700-900 lines/s (down); 700-100 lines/s (up)

So, in these tests, this PR makes no significant difference.

However, scrolling with the mouse scrollwheel over a separate editor panel in which the textcursor is not focused seems to be necessary to trigger the main behavior as @mrclary and others have reported, as what was reported is that the major slowdown only happens in Editor panels in which the textcursor is not focused, which is necessary for keyboard-based scrolling. 

Indeed, in my qualitative testing, I was able to apparently repro this: scrolling is substantially smoother in Editor panels that have focus on 4.x and quite jerky on others, while it is equally smooth on both using this PR. I can try to script this with AHK, but it appears most of the problem is with laggy updates and jerky movement, which is much harder to measure quantitatively than simply speed (though I can try to measure the lags if they're long enough, it will be pretty imprecise).

I could record my screen and measure update framerate with various settings, but that would be pretty time-intensive for uncertain gains. Given scrolling is already very smooth for me even in a large file (mainwindow.py, 3700 lines) and I don't see any perceptible lag in decoration updates, not sure its worth tuning much further but I would appreciate the feedback of other users as well.

**Review by CAM-Gerlach — APPROVED** [review]
Per my testing, this does appear to resolve the problem without perceptible lag either scrolling or updating decorations. Great work @ccordoba12 , thanks!

I didn't review it in line by line detail, but the code looked good overall. One minor note: `_` is used as a dummy variable, but it is mapped to `gettext` elsewhere, so by our convention and to avoid any confusion or spurious grep/find in files results for it, I suggest using the standard `__` instead (there were 5 instances, all in the unit test file).

**Comment by ccordoba12** [comment]
> One minor note: _ is used as a dummy variable, but it is mapped to gettext elsewhere, so by our convention and to avoid any confusion or spurious grep/find in files results for it, I suggest using the standard __ instead (there were 5 instances, all in the unit test file).

Good point. I addressed that in my last commit.

File: spyder/plugins/editor/widgets/codeeditor.py
Comment (dalthviz) [inline spyder/plugins/editor/widgets/codeeditor.py:3845]:
```suggestion
        # See https://bugreports.qt.io/browse/QTBUG-25365
```

**Review by dalthviz — COMMENTED** [review]
Thanks @ccordoba12 I left a simple suggestion otherwise this LGTM :+1:

Also, I don't know if someone else wants to manually check this to provide more feedback before the merge. Probably by the end of the day I will try to merge this if the suggestion is applied.

**Comment by dalthviz** [comment]
Oh and also @ccordoba12, seems like a PEP8 issue remains:

> Hello @ccordoba12! Thanks for updating this PR. We checked the lines you've touched for PEP 8 issues, and found:
> 
>     In the file spyder/plugins/editor/utils/decoration.py:
> 
>     Line 29:20: E261 at least two spaces before inline comment
> 
> Comment last updated at 2020-09-22 17:36:37 UTC

**Comment by ccordoba12** [comment]
> Oh and also @ccordoba12, seems like a PEP8 issue remains

Good eye! Fixed in my last commit.

**Review by dalthviz — APPROVED** [review]
Thanks @ccordoba12 LGTM :+1:

Also, thank you all for the help reviewing this and the feedback :)

**Pull request merged**
